ext-apiserver/gcp-deployer/deploy/deploy_helper.go

```diff
@@ -90,76 +85,48 @@ func (d *deployer) createCluster(c *clusterv1.Cluster, machines []*clusterv1.Mac
 	if err := d.initApiClient(); err != nil {
 		return err
 	}
-	glog.Info("Starting the machine controller...")
-	//if err := d.machineDeployer.CreateMachineController(c, machines); err != nil {
-	//	return fmt.Errorf("can't create machine controller: %v", err)
-	//}
-	//
-	//if err := d.createClusterCRD(); err != nil {
-	//	return err
-	//}
-
-	if _, err := d.client.Clusters(apiv1.NamespaceDefault).Create(c); err != nil {
+
+	glog.Info("Deploying the addon apiserver and controller manager...")
+	if err := d.machineDeployer.CreateMachineController(c, machines); err != nil {
+		return fmt.Errorf("can't create machine controller: %v", err)
+	}
+
+	if err := d.waitForClusterResourceReady(); err != nil {
 		return err
 	}
 
-	//if err := d.createMachineCRD(); err != nil {
-	//	return err
-	//}
+	c, err = d.client.Clusters(apiv1.NamespaceDefault).Create(c)
+	if err != nil {
+		return err
+	} else {
+		c.Status.APIEndpoints = append(c.Status.APIEndpoints,
+			clusterv1.APIEndpoint{
+				Host: masterIP,
+				Port: 443,
+			})
+		if _, err := d.client.Clusters(apiv1.NamespaceDefault).UpdateStatus(c); err != nil {
+			return err
+		}
+	}
 
 	if err := d.createMachines(machines); err != nil {
 		return err
 	}
 	return nil
 }
 
-//func (d *deployer) createClusterCRD() error {
-//	cs, err := util.NewClientSet(d.configPath)
-//	if err != nil {
-//		return err
-//	}
-//
-//	success := false
-//	for i := 0; i <= RetryAttempts; i++ {
-//		if _, err = clusterv1.CreateClustersCRD(cs); err != nil {
-//			glog.Info("Failure creating Clusters CRD (will retry).")
-//			time.Sleep(SleepSecondsPerAttempt * time.Second)
-//			continue
-//		}
-//		success = true
-//		glog.Info("Clusters CRD created succuessfully!")
-//		break
-//	}
-//
-//	if !success {
-//		return fmt.Errorf("error creating Clusters CRD: %v", err)
-//	}
-//	return nil
-//}
-//
-//func (d *deployer) createMachineCRD() error {
-//	cs, err := util.NewClientSet(d.configPath)
-//	if err != nil {
-//		return err
-//	}
-//
-//	success := false
-//	for i := 0; i <= RetryAttempts; i++ {
-//		if _, err = clusterv1.CreateMachinesCRD(cs); err != nil {
-//			glog.Info("Failure creating Machines CRD (will retry).")
-//			time.Sleep(time.Duration(SleepSecondsPerAttempt) * time.Second)
-//			continue
-//		}
-//		success = true
-//		glog.Info("Machines CRD created successfully!")
-//		break
-//	}
-//
-//	if !success {
-//		return fmt.Errorf("error creating Machines CRD: %v", err)
-//	}
-//	return nil
-//}
+func (d *deployer) waitForClusterResourceReady() error {
+	err := wait.Poll(500*time.Millisecond, 120*time.Second, func() (bool, error) {
+		_, err := d.client.Clusters(apiv1.NamespaceDefault).List(metav1.ListOptions{})
```

Comment:
This should use the discovery API instead and examine the published APIs through that.

Reply:
Did you miss this comment?